src/halmos/__main__.py

```diff
@@ -397,6 +408,279 @@ def is_global_fail_set(context: CallContext) -> bool:
+        # update `exs` with the new output states, and `funsigs` with remaining invariant tests for the next depth.
+        exs, funsigs = step_invariant_tests(inv_ctx, exs, funsigs, depth)
+
+        # TODO: merge, simplify, or prioritize exs to mitigate path explosion
+
+        if args.debug:
+            print(f"{depth=}\n")
+            for idx, ex in enumerate(exs):
+                print(f"{idx=} {hexify(get_state_id(ex))=}\n")
+                print(ex)
+                render_trace(ex.context)
+
+        # stop if no new frontier states or remaining invariant tests
+        if not exs or not funsigs:
+            break
+
+    test_results = test_results_map.values()
+
+    # print passed tests; failed tests have already been displayed during step_invariant_tests()
+    for r in test_results:
+        if r.exitcode == PASS:
+            print(
+                f"{green('[PASS]')} {r.name} (depth: {depth-1}, paths: {len(visited)})"
+            )
+
+    return test_results
+
+
+def step_invariant_tests(
+    inv_ctx: InvariantContext,
+    pre_exs: list[Exec],
+    funsigs: list[str],
+    depth: int,
+) -> tuple[list[Exec], list[str]]:
+    """
+    Executes the next depth of the invariant testing run, which consists of:
+    1. Computing new frontier states by executing an arbitrary function of an arbitrary target contract from each of the given input states.
+    2. Executing invariant tests for each new frontier state.
+    3. Returning the new frontier states that have not been visited.
+
+    Args:
+        inv_ctx: The context for the invariant testing run.
+        pre_exs: A list of input states.
+        funsigs: A list of invariant test function signatures to be executed at this depth.
+        depth: The current depth of the invariant testing run.
+
+    Returns:
+        A list of new frontier states generated at this depth.
+        A list of invariant test function signatures remaining for the next depth.
+    """
+    ctx = inv_ctx.contract_ctx
+    test_results_map = inv_ctx.test_results_map
+    visited = inv_ctx.visited
+
+    next_exs = []
+
+    for idx, pre_ex in enumerate(pre_exs):
+        progress_status.update(f"{depth=} {len(visited)=} {idx=} {len(pre_exs)=}")
+
+        for addr in pre_ex.code:
+            # skip the test contract
+            if eq(addr, con_addr(FOUNDRY_TEST)):
+                continue
+
+            # execute a target contract
+            post_exs = run_target_contract(ctx, pre_ex, addr)
+
+            for post_ex in post_exs:
+                subcall = post_ex.context
+
+                # ignore and report if halmos-errored
+                if subcall.is_stuck():
+                    error(
+                        f"{depth=}: addr={hexify(addr)}: {subcall.get_stuck_reason()}"
+                    )
+                    continue
+
+                # ignore if reverted
+                if subcall.output.error:
+                    continue
```

Comment:
what if it reverted with an assertion failure?

Reply:
it's for target contracts, and we don't expect assert cheatcodes there.